src/libsystemd/sd-bus/bus-track.c

```diff
@@ -191,12 +191,12 @@ _public_ int sd_bus_track_add_name(sd_bus_track *track, const char *name) {
         i = hashmap_get(track->names, name);
         if (i) {
                 if (track->recursive) {
-                        unsigned k = track->n_ref + 1;
+                        unsigned k = i->n_ref + 1;
 
-                        if (k < track->n_ref) /* Check for overflow */
+                        if (k < i->n_ref) /* Check for overflow */
```

Comment:
we should not do overflow checks this way really. We used to, but it's not in compliance with C standard. i.e. overflows are undefined behaviour in C, and thus the compiler may assume that the the earlier +1 doesn't overflow, and thus optimize this check away.

thus, this should be reworked to:

```c
if (i->n_ref >= UINT_MAX)
        return -EOVEFLOW;

i->n_ref++;
```

i.e. never execute the +1 unless we verified it's safe.

Reply:
Sure, but this code block will be replaced with `track_item_ref()` in the later commit. I'd like to keep the commit as is.

Reply:
Given that `sd_bus_track_add_name` is public I wonder if it makes sense to mention in NEWS that it's going to either crash or overflow `i->n_ref` with NDEBUG defined if its callers manage to trigger it enough times. FWIW I don't think PID1 should call functions that can crash it like that.

Reply:
People shouldn't set NDEBUG. That's generally a bad idea. That said, the overflow assert here should be upgraded to assert_se(), as it should never be optimized away. I filed #23138 for that.

And yes, I guess given that this is directly user controllable, we should not use the generic ref counting code here (where I think the assertion is OK, as there shoud be limits on resources by other means anyway), nd instead have manual one that returns an error instead of an assertion back.

i.e. assertions are OK for local programming errors, and in general local ref counting overflows are that: objects simply should have been limited in other ways beforehand. But this here is not a local ref counting logic, but a directly remotely accessible in the usual ways, hence we should return an error here.

Reply:
→ #23139